crates/matrix-sdk-crypto/src/identities/manager.rs

```diff
@@ -414,151 +418,227 @@ impl IdentityManager {
-                        "Couldn't update or create new user identity"
+                        expected = ?self.user_id(),
+                        got = ?user_signing.user_id(),
+                        "User ID mismatch in our user-signing key",
                     );
-                    continue;
+                    Err(SignatureError::UserIdMismatch)
+                } else {
+                    let identity =
+                        ReadOnlyOwnUserIdentity::new(master_key, self_signing, user_signing)?;
+
+                    let private = self.check_private_identity(&identity).await;
+
+                    Ok(IdentityChange { public: identity.into(), private })
+                }
+            } else {
+                warn!(
+                    "User identity for our own user didn't contain a user signing pubkey or the key \
+                    isn't valid",
+                );
+
+                Err(SignatureError::MissingSigningKey)
+            }
+        } else {
+            let identity = ReadOnlyUserIdentity::new(master_key, self_signing)?;
+            Ok(IdentityChange { public: identity.into(), private: None })
+        }
+    }
+
+    /// Try to deserialize the the master key and self-signing key of a
+    /// identity.
+    ///
+    /// Each user identity *must* at least contain a master and self-signing
+    /// key. Our own identity, in addition to those two, also contains a
+    /// user-signing key.
+    fn get_minimal_set_of_keys(
+        master_key: &Raw<CrossSigningKey>,
+        response: &KeysQueryResponse,
+    ) -> Option<(MasterPubkey, SelfSigningPubkey)> {
+        match master_key.deserialize_as::<MasterPubkey>() {
+            Ok(master_key) => {
+                if let Some(self_signing) = response
+                    .self_signing_keys
+                    .get(master_key.user_id())
+                    .and_then(|k| k.deserialize_as::<SelfSigningPubkey>().ok())
+                {
+                    Some((master_key, self_signing))
+                } else {
+                    warn!("A user identity didn't contain a self signing pubkey or the key was invalid");
+                    None
+                }
+            }
+            Err(e) => {
+                warn!(
+                    error = ?e,
+                    "Couldn't update or create new user identity"
+                );
+                None
+            }
+        }
+    }
+
+    #[instrument(skip_all, fields(user_id))]
+    async fn update_or_create_identity(
+        &self,
+        response: &KeysQueryResponse,
+        changes: &mut IdentityChanges,
+        changed_identity: &mut Option<PrivateCrossSigningIdentity>,
+        user_id: &UserId,
+        master_key: MasterPubkey,
+        self_signing: SelfSigningPubkey,
+    ) -> StoreResult<()> {
+        if master_key.user_id() != user_id || self_signing.user_id() != user_id {
+            warn!(?user_id, "User ID mismatch in one of the cross signing keys",);
```

Comment:
Shouldn't we return an `Err` in this case? The inputs are incorrect, so I bet it's an error?

Reply:
We can't return an error, a `/keys/query` response will contain many identities, the identities are provided by other clients and may by the nature of user-provided data be completely invalid.

In other words, we can't discard the whole response because a single buggy or malicious client.